[R-package] 32-bit library support

## Summary

As I've investigated #2960 as part of #629 , I've learned that packages on CRAN have to be able to build on 32-bit Windows.

**I believe that we need to add support for a 32-bit build LightGBM to get the R package to CRAN.**

### Evidence that this is required

I have not yet found any official statement that says "all R packages have to build on 32-bit Windows", but I see strong evidence that that is true.

You can find the current list of check environments at https://cran.r-project.org/web/checks/check_flavors.html. None of those seem to be 32-bit

![image](https://user-images.githubusercontent.com/7608904/85821730-e8371200-b73e-11ea-9f3d-29cb21c9e187.png)

but in ["Writing R Extensions"](https://cran.r-project.org/doc/manuals/r-release/R-exts.html) I see the following:

> On systems which support multiple sub-architectures (principally Windows), R CMD check will install and check a package which contains compiled code under all available sub-architectures.

And in the ["R  for Windows FAQs"](https://cran.r-project.org/bin/windows/base/rw-FAQ.html):

> Packages without compiled code nor a configure.win script will run on both 32- and 64-bit R.

> Packages with compiled code but no configure.win nor src/Makefile.win file will be built for both when running on a 64-bit version of Windows if both versions of R are installed.

> An empty configure.win is treated in the same way as if none existed. Also, there is a list of packages known to have an architecture-independent configure.win hardcoded into R CMD INSTALL, and for these packages, both architectures will be built under the above conditions. Other packages can be installed with configure.win run for just the first architecture by using option --force-biarch.

> Any package can be installed for first one architecture and then the other with option --merge-multiarch, but the package source must be a tarball (and as before, running on a 64-bit version of Windows with both versions of R installed).

> Finally, a package without a src/Makefile.win file and no or empty or architecture-independent configure.win file can be installed for both architectures from 32-bit Windows if the 64-bit components were selected when R was installed and option --compile-both is given. Obviously, only the 32-bit installation can be tested.

This is why you see pull requests this in other projects: https://github.com/dmlc/xgboost/pull/5777

I've tried looking in [the source for R CMD check](https://github.com/wch/r-source/blob/f27cbf1a52a31cd9b9676340394946a22041a4ae/src/library/tools/R/check.R) but it is very difficult to follow.

## Motivation

I believe that we need to add support for a 32-bit build LightGBM to get the R package to CRAN.

## Description

To check if the R package works with 32-bit R, try the following:

```shell
sh build-cran-package.sh
"C:\Program Files\R\R-4.0.2\bin\i386\R.exe" CMD check lightgbm_3.0.0-1.tar.gz --as-cran
```

## References

There was some discussion about this in #1129 , where it was decided that supporting a 32-bit is possible but that the maintenance burden would be too much. Given that it is needed to get LightGBM to CRAN, I think it should be reconsidered.

> I don't think so. I think the only two R Hub 32-bit Linux builds were on Solaris and those both failed (#629 (comment))

I saw the errors in 32-bit version are caused by `network/socket_wrapper.hpp:30:21: fatal error: ifaddrs.h: No such file or directory` . And the windows bulid for both 32-bit and 64-bit seems all pass.

Ok I just tried with MSVC. You can see the changes I had to make here: https://github.com/microsoft/LightGBM/compare/master...jameslamb:r/try-32bit-msvc?expand=1.

I ran this

```shell
"C:\Program Files\R\R-4.0.2\bin\i386\Rscript.exe" build_r.R
```

Here's as much log as I was able to copy:

<details><summary>build_r.R logs</summary>

```text
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\boosting\score_updater.hpp(32,46): warning C4244: 'argument': conversion from 'int64_t' to 'size_t', possible loss of data (compiling source file C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\boosting\gbdt_model_text.cpp) [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\boosting\score_updater.hpp(44,16): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data (compiling source file C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\boosting\gbdt_model_text.cpp) [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\boosting\goss.hpp(66,20): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data (compiling source file C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\boosting\boosting.cpp) [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\boosting\goss.hpp(67,19): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data (compiling source file C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\boosting\boosting.cpp) [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
  gbdt_prediction.cpp
  prediction_early_stop.cpp
  bin.cpp
  config.cpp
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\boosting\score_updater.hpp(30,19): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data (compiling source file C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\boosting\gbdt_prediction.cpp) [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\boosting\score_updater.hpp(32,46): warning C4244: 'argument': conversion from 'int64_t' to 'size_t', possible loss of data (compiling source file C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\boosting\gbdt_prediction.cpp) [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\boosting\score_updater.hpp(44,16): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data (compiling source file C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\boosting\gbdt_prediction.cpp) [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
  config_auto.cpp
  dataset.cpp
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\io\dataset.cpp(635,33): warning C4244: 'argument': conversion from 'uint64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
  dataset_loader.cpp
  file_io.cpp
  json11.cpp
  metadata.cpp
  parser.cpp
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\io\metadata.cpp(265,43): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\io\metadata.cpp(298,17): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\io\metadata.cpp(407,37): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
  tree.cpp
  dcg_calculator.cpp
  metric.cpp
  linker_topo.cpp
  linkers_mpi.cpp
  linkers_socket.cpp
  network.cpp
  objective_function.cpp
  data_parallel_tree_learner.cpp
  feature_parallel_tree_learner.cpp
  gpu_tree_learner.cpp
  serial_tree_learner.cpp
  tree_learner.cpp
  voting_parallel_tree_learner.cpp
  c_api.cpp
  lightgbm_R.cpp
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(463,11): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(464,37): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(471,29): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(480,29): warning C4244: 'initializing': conversion from 'int64_t' to 'unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(483,30): warning C4244: 'initializing': conversion from 'int64_t' to 'unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(488,32): warning C4244: 'initializing': conversion from 'int64_t' to 'unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(503,33): warning C4244: 'initializing': conversion from 'int64_t' to 'unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(506,33): warning C4244: 'initializing': conversion from 'int64_t' to 'unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(512,67): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(516,45): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(517,58): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(521,31): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(524,19): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(526,30): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(528,30): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(528,94): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(528,135): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(547,31): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(549,52): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(560,51): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(560,80): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(599,67): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(609,31): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(633,33): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(633,56): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(633,89): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(635,116): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(637,95): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(642,65): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(643,50): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(657,31): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(660,59): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(662,30): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(666,28): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(1134,52): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(1135,46): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(1200,52): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(1201,46): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(1271,61): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(1272,55): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(2188,41): warning C4244: 'argument': conversion from 'int64_t' to 'size_t', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(2206,41): warning C4244: 'argument': conversion from 'int64_t' to 'size_t', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(2391,27): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(979): message : see reference to function template instantiation 'std::function<std::vector<std::pair<int,double>,std::allocator<std::pair<int,double>>> (int)> RowFunctionFromCSR<int>(const void *,int,const int32_t *,const void *,int,int64_t,int64_t)' being compiled [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(2405,27): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(2422,27): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\src\c_api.cpp(2436,27): warning C4244: 'argument': conversion from 'int64_t' to 'const unsigned int', possible loss of data [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
     Creating library C:/Users/James/AppData/Local/Temp/RtmpikVs2F/R.INSTALL3ba03d177120/lightgbm/src/Release/lib_lightgbm.lib and object C:/Users/James/AppData/Local/Temp/RtmpikVs2F/R.INSTALL3ba03d177120/lightgbm/src/Release/lib_lightgbm.exp
c_api.obj : error LNK2001: unresolved external symbol _Rf_error [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
lightgbm_R.obj : error LNK2001: unresolved external symbol _Rf_error [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
feature_parallel_tree_learner.obj : error LNK2001: unresolved external symbol _Rf_error [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
serial_tree_learner.obj : error LNK2001: unresolved external symbol _Rf_error [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
tree_learner.obj : error LNK2001: unresolved external symbol _Rf_error [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
voting_parallel_tree_learner.obj : error LNK2001: unresolved external symbol _Rf_error [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
linkers_socket.obj : error LNK2001: unresolved external symbol _Rf_error [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
network.obj : error LNK2001: unresolved external symbol _Rf_error [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
objective_function.obj : error LNK2001: unresolved external symbol _Rf_error [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
data_parallel_tree_learner.obj : error LNK2001: unresolved external symbol _Rf_error [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
parser.obj : error LNK2001: unresolved external symbol _Rf_error [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
tree.obj : error LNK2001: unresolved external symbol _Rf_error [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
dcg_calculator.obj : error LNK2001: unresolved external symbol _Rf_error [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
metric.obj : error LNK2001: unresolved external symbol _Rf_error [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
dataset.obj : error LNK2001: unresolved external symbol _Rf_error [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
dataset_loader.obj : error LNK2001: unresolved external symbol _Rf_error [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
json11.obj : error LNK2001: unresolved external symbol _Rf_error [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
metadata.obj : error LNK2001: unresolved external symbol _Rf_error [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
prediction_early_stop.obj : error LNK2001: unresolved external symbol _Rf_error [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
bin.obj : error LNK2001: unresolved external symbol _Rf_error [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
config.obj : error LNK2001: unresolved external symbol _Rf_error [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
config_auto.obj : error LNK2001: unresolved external symbol _Rf_error [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
application.obj : error LNK2001: unresolved external symbol _Rf_error [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
boosting.obj : error LNK2001: unresolved exterError in .run_shell_command(build_cmd, build_args) :
  Command failed with exit code: 1
nal symbol _Rf_error [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
gbdt.obj : error LNK2001: unresolved external symbol _Rf_error [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
gbdt_model_text.obj : error LNK2001: unresolved external symbol _Rf_error [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
serial_tree_learner.obj : error LNK2001: unresolved external symbol _Rprintf [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
voting_parallel_tree_learner.obj : error LNK2001: unresolved external symbol _Rprintf [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
c_api.obj : error LNK2001: unresolved external symbol _Rprintf [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
linkers_socket.obj : error LNK2001: unresolved external symbol _Rprintf [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
network.obj : error LNK2001: unresolved external symbol _Rprintf [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
objective_function.obj : error LNK2001: unresolved external symbol _Rprintf [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
data_parallel_tree_learner.obj : error LNK2001: unresolved external symbol _Rprintf [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
metadata.obj : error LNK2001: unresolved external symbol _Rprintf [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
parser.obj : error LNK2001: unresolved external symbol _Rprintf [* removing 'C:/Users/James/Documents/R/win-library/4.0/lightgbm'
* restoring previous 'C:/Users/James/Documents/R/win-library/4.0/lightgbm'
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
tree.obj : error LNK2001: unresolved external symbol _Rprintf [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
metric.obj : error LNK2001: unresolved external symbol _Rprintf [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
bin.obj : error LNK2001: unresolved external symbol _Rprintf [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
config.obj : error LNK2001: unresolved external symbol _Rprintf [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
dataset.obj : error LNK2001: unresolved external symbol _Rprintf [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
dataset_loader.obj : error LNK2001: unresolved external symbol _Rprintf [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
application.obj : error LNK2001: unresolved external symbol _Rprintf [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
boosting.obj : error LNK2001: unresolved external symbol _Rprintf [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
gbdt.obj : error LNK2001: unresolved external symbol _Rprintf [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
gbdt_model_text.obj : error LNK2001: unresolved external symbol _Rprintf [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
serial_tree_learner.obj : error LNK2001: unresolved external symbol _Rvprintf [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
voting_parallel_tree_learner.obj : error LNK2001: unresolved external symbol _Rvprintf [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
c_api.obj : error LNK2001: unresolved external symbol _Rvprintf [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
linkers_socket.obj : error LNK2001: unresolved external symbol _Rvprintf [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
network.obj : error LNK2001: unresolved external symbol _Rvprintf [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
objective_function.obj : error LNK2001: unresolved external symbol _Rvprintf [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
data_parallel_tree_learner.obj : error LNK2001: unresolved external symbol _Rvprintf [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
metadata.obj : error LNK2001: unresolved external symbol _Rvprintf [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
parser.obj : error LNK2001: unresolved external symbol _Rvprintf [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
tree.obj : error LNK2001: unresolved external symbol _Rvprintf [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
metric.obj : error LNK2001: unresolved external symbol _Rvprintf [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
bin.obj : error LNK2001: unresolved external symbol _Rvprintf [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
config.obj : error LNK2001: unresolved external symbol _Rvprintf [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
dataset.obj : error LNK2001: unresolved external symbol _Rvprintf [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
dataset_loader.obj : error LNK2001: unresolved external symbol _Rvprintf [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
application.obj : error LNK2001: unresolved external symbol _Rvprintf [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
boosting.obj : error LNK2001: unresolved external symbol _Rvprintf [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
gbdt.obj : error LNK2001: unresolved external symbol _Rvprintf [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
gbdt_model_text.obj : error LNK2001: unresolved external symbol _Rvprintf [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
lightgbm_R.obj : error LNK2019: unresolved external symbol _R_registerRoutines referenced in function "void __cdecl R_init_lightgbm(struct _DllInfo *)" (?R_init_lightgbm@@YAXPAU_DllInfo@@@Z) [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
lightgbm_R.obj : error LNK2019: unresolved external symbol _R_useDynamicSymbols referenced in function "void __cdecl R_init_lightgbm(struct _DllInfo *)" (?R_init_lightgbm@@YAXPAU_DllInfo@@@Z) [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
R.lib : warning LNK4272: library machine type 'x64' conflicts with target machine type 'x86' [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\Release\lib_lightgbm.dll : fatal error LNK1120: 5 unresolved externals [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
Error in .run_shell_command(install_cmd, install_args) :
  Command failed with exit code: 1
Execution halted
```

</details>

There were a lot of warnings and linker errors. here's a small sample

> [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
lightgbm_R.obj : error LNK2019: unresolved external symbol _R_useDynamicSymbols referenced in function "void __cdecl R_init_lightgbm(struct _DllInfo *)" (?R_init_lightgbm@@YAXPAU_DllInfo@@@Z) [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
R.lib : warning LNK4272: library machine type 'x64' conflicts with target machine type 'x86' [C:\Users\James\AppData\Local\Temp\RtmpikVs2F\R.INSTALL3ba03d177120\lightgbm\src\build\_lightgbm.vcxproj]
C:\Users\James\AppData\Local\Temp\Rtm

@guolinke 

> I saw the errors in 32-bit version are caused by `network/socket_wrapper.hpp:30:21: fatal error: ifaddrs.h: No such file or directory` . 

What do you think about the possible solution I proposed with borrowing `ifaddrs.h` for Solaris?
https://github.com/microsoft/LightGBM/issues/629#issuecomment-665091451

--- a/R-package/R/lgb.Booster.R
+++ b/R-package/R/lgb.Booster.R
@@ -31,7 +31,7 @@ Booster <- R6::R6Class(
 
       # Create parameters and handle
       params <- append(params, list(...))
-      handle <- 0.0
+      handle <- 0L
 
       # Attempts to create a handle for the dataset
       try({

--- a/R-package/R/lgb.Booster.R
+++ b/R-package/R/lgb.Booster.R
@@ -31,7 +31,7 @@ Booster <- R6::R6Class(
 
       # Create parameters and handle
       params <- append(params, list(...))
-      handle <- 0L
+      handle <- NA_integer_
 
       # Attempts to create a handle for the dataset
       try({

--- a/R-package/R/lgb.Booster.R
+++ b/R-package/R/lgb.Booster.R
@@ -31,7 +31,7 @@ Booster <- R6::R6Class(
 
       # Create parameters and handle
       params <- append(params, list(...))
-      handle <- NA_integer_
+      handle <- lgb.null.handle()
 
       # Attempts to create a handle for the dataset
       try({

--- a/R-package/R/lgb.Dataset.R
+++ b/R-package/R/lgb.Dataset.R
@@ -186,7 +186,7 @@ Dataset <- R6::R6Class(
       if (!is.null(private$reference)) {
         ref_handle <- private$reference$.__enclos_env__$private$get_handle()
       }
-      handle <- 0L
+      handle <- NA_integer_
 
       # Not subsetting
       if (is.null(private$used_indices)) {

--- a/R-package/R/lgb.Dataset.R
+++ b/R-package/R/lgb.Dataset.R
@@ -186,7 +186,7 @@ Dataset <- R6::R6Class(
       if (!is.null(private$reference)) {
         ref_handle <- private$reference$.__enclos_env__$private$get_handle()
       }
-      handle <- NA_real_
+      handle <- 0L
 
       # Not subsetting
       if (is.null(private$used_indices)) {

--- a/R-package/R/lgb.Dataset.R
+++ b/R-package/R/lgb.Dataset.R
@@ -186,7 +186,7 @@ Dataset <- R6::R6Class(
       if (!is.null(private$reference)) {
         ref_handle <- private$reference$.__enclos_env__$private$get_handle()
       }
-      handle <- NA_integer_
+      handle <- lgb.null.handle()
 
       # Not subsetting
       if (is.null(private$used_indices)) {

--- a/R-package/R/utils.R
+++ b/R-package/R/utils.R
@@ -6,6 +6,14 @@ lgb.is.Dataset <- function(x) {
   lgb.check.r6.class(x, "lgb.Dataset") # Checking if it is of class lgb.Dataset or not
 }
 
+lgb.null.handle <- function() {
+  if (.Machine$sizeof.pointer == 8L) {
+    return(NA_real_)
+  } else {
+    return(NA_integer_)
+  }
+}
+
 lgb.is.null.handle <- function(x) {
   is.null(x) || is.na(x)
 }

--- a/R-package/tests/testthat/test_dataset.R
+++ b/R-package/tests/testthat/test_dataset.R
@@ -73,7 +73,7 @@ test_that("lgb.Dataset: nrow is correct for a very sparse matrix", {
 
 test_that("lgb.Dataset: Dataset should be able to construct from matrix and return non-null handle", {
   rawData <- matrix(runif(1000L), ncol = 10L)
-  handle <- 0L
+  handle <- NA_integer_
   ref_handle <- NULL
   handle <- lightgbm:::lgb.call(
     "LGBM_DatasetCreateFromMat_R"

--- a/R-package/tests/testthat/test_dataset.R
+++ b/R-package/tests/testthat/test_dataset.R
@@ -73,7 +73,7 @@ test_that("lgb.Dataset: nrow is correct for a very sparse matrix", {
 
 test_that("lgb.Dataset: Dataset should be able to construct from matrix and return non-null handle", {
   rawData <- matrix(runif(1000L), ncol = 10L)
-  handle <- NA_integer_
+  handle <- lgb.null.handle()
   ref_handle <- NULL
   handle <- lightgbm:::lgb.call(
     "LGBM_DatasetCreateFromMat_R"

--- a/R-package/tests/testthat/test_dataset.R
+++ b/R-package/tests/testthat/test_dataset.R
@@ -73,7 +73,7 @@ test_that("lgb.Dataset: nrow is correct for a very sparse matrix", {
 
 test_that("lgb.Dataset: Dataset should be able to construct from matrix and return non-null handle", {
   rawData <- matrix(runif(1000L), ncol = 10L)
-  handle <- NA_real_
+  handle <- 0L
   ref_handle <- NULL
   handle <- lightgbm:::lgb.call(
     "LGBM_DatasetCreateFromMat_R"

